packages/smithy-http/src/smithy_http/interceptors/user_agent.py

```diff
@@ -0,0 +1,151 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  SPDX-License-Identifier: Apache-2.0
+import platform
+from typing import Self, Any
+
+import smithy_core
+from smithy_core.interceptors import Interceptor, InterceptorContext
+from smithy_http import Field
+from smithy_http.aio.interfaces import HTTPRequest
+from smithy_http.user_agent import UserAgent, UserAgentComponent
+
+
+class UserAgentInterceptor(Interceptor[Any, None, HTTPRequest, None]):
+    """Adds interceptors that initialize UserAgent in the context and add the user-agent
+    header."""
+
+    def read_before_execution(
+        self, context: InterceptorContext[Any, None, None, None]
+    ) -> None:
+        context.properties["user_agent"] = UserAgentBuilder.from_environment().build()
+
+    def modify_before_signing(
+        self, context: InterceptorContext[Any, None, HTTPRequest, None]
+    ) -> HTTPRequest:
+        user_agent = context.properties["user_agent"]
+        request = context.transport_request
+        request.fields.set_field(Field(name="User-Agent", values=[str(user_agent)]))
+        return context.transport_request
+
+
+_USERAGENT_ALLOWED_OS_NAMES = (
+    "windows",
+    "linux",
+    "macos",
+    "android",
+    "ios",
+    "watchos",
+    "tvos",
+    "other",
+)
+_USERAGENT_PLATFORM_NAME_MAPPINGS = {"darwin": "macos"}
+_USERAGENT_SDK_NAME = "python"
+
+
+class UserAgentBuilder:
+    def __init__(
+        self,
+        *,
+        platform_name: str | None = None,
+        platform_version: str | None = None,
+        platform_machine: str | None = None,
+        python_version: str | None = None,
+        python_implementation: str | None = None,
```

Comment:
I think the initial idea of having all of these required but nullable is to force the implementer to make a decision about information gathering. If you forget something it throws an error. We can make it optional if there are usability concerns, just want to make sure we're consciously making that tradeoff.

Reply:
It seems generically useful to be able to construct with limited information... however, I think we shouldn't consider this as a generic class.  I've updated it to private again and I'll remove these defaults to force initialization with all values.